cmd/calico-ipam-upgrade/main.go

```diff
@@ -0,0 +1,298 @@
+// Copyright 2015 Tigera Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package main
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path"
+	"strings"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/projectcalico/libcalico-go/lib/apiconfig"
+	bapi "github.com/projectcalico/libcalico-go/lib/backend/api"
+	"github.com/projectcalico/libcalico-go/lib/backend/k8s"
+	client "github.com/projectcalico/libcalico-go/lib/clientv3"
+	"github.com/projectcalico/libcalico-go/lib/errors"
+	"github.com/projectcalico/libcalico-go/lib/ipam"
+	"github.com/projectcalico/libcalico-go/lib/logutils"
+	"github.com/projectcalico/libcalico-go/lib/net"
+	cnet "github.com/projectcalico/libcalico-go/lib/net"
+	"github.com/projectcalico/libcalico-go/lib/options"
+
+	"github.com/projectcalico/cni-plugin/internal/pkg/utils"
+)
+
+const (
+	ipAllocPath = "/var/lib/cni/networks/k8s-pod-network"
+)
+
+var (
+	binariesToDisable = []string{
+		"/host/opt/cni/bin/calico",
+		"/host/opt/cni/bin/host-local",
+	}
+)
+
+type Accessor interface {
+	Backend() bapi.Client
+}
+
+func main() {
+	log.Info("migrating from host-local to calico-ipam...")
+	ctxt := context.Background()
+
+	// nodename associates IPs to this node.
+	nodename := os.Getenv("KUBERNETES_NODE_NAME")
+	logCtxt := log.WithField("node", nodename)
+
+	// calicoClient makes IPAM calls.
+	cfg, err := apiconfig.LoadClientConfig("")
+	if err != nil {
+		logCtxt.Fatal("failed to load api client config")
+	}
+	cfg.Spec.DatastoreType = apiconfig.Kubernetes
+	calicoClient, err := client.New(*cfg)
+	if err != nil {
+		logCtxt.Fatal("failed to initialize api client")
+	}
+
+	// Execute migration
+	if err := Migrate(ctxt, calicoClient, nodename); err != nil {
+		logCtxt.WithError(err).Fatal("failed to migrate ipam")
+	}
+}
+
+func Migrate(ctxt context.Context, c client.Interface, nodename string) error {
+	// Set up logging formatting.
+	log.SetFormatter(&logutils.Formatter{})
+
+	// Install a hook that adds file/line no information.
+	log.AddHook(&logutils.ContextHook{})
+
+	// k8sClient directly calls the k8s apiserver.
+	k8sClient := c.(Accessor).Backend().(*k8s.KubeClient).ClientSet
+
+	var err error
+	defer func() {
```

Comment:
What's this for?

Reply:
This is the best approach I could think to give the user time to view the error in the logs if one occurs without it getting overwritten by another error in a future execution.

Reply:
Rather than the defer, we should at least put it in the error handling when we call `Migrate` above. I think that's easier to reason about.

But, I'm not sure a sleep is really the best thing here - why don't we just implement the retry logic ourselves? What is the value in having the init container exit?

e.g.

```go
for {
  err := Migrate()
  if err != nil {
    // log error
    time.Sleep(1 * time.Second)
    continue
  }
}
```